### ISSUE TITLE
Circular Color Picker: Improve HTML Semantics and Interactions

### ISSUE THREAD

**Issue by jeryj** [WordPress/gutenberg#20372]
The Circular Color Picker is currently marked up as grouping of `<div>`s.

**To reproduce**
1. Go to the circular color picker
2. Tab moves to the next color

**Expected behavior**
1. Go to the circular color picker
2. Right/Left/Up/Down should move to the relevant next/previous items
3. Tab moves to the next tabbable item (like the custom color button)

**Screenshots**
<img width="259" alt="Circular Color Picker - Group of colored circles with a custom color link/button and clear button" src="https://user-images.githubusercontent.com/967608/75075413-6058a700-54c3-11ea-8f99-aa950860c5f3.png">

**Proposal to be discussed**
I think the ideal is to convert this to a `<fieldset>` with a `<legend>` and group of radio buttons (styled to look identical to the current implementation). This HTML structure would be semantic, and give us all the expected keyboard interactions. Because it's ideal that each radio grouping has at least one selection, we could add in a "Custom Color" circular option as well that is the value of whatever is in the color selector.

If this isn't possible, then we could convert this to a grouping of buttons using an `<ul>` and each color `<button>` being inside the `<li>`.  There may be better `aria` states to communicate this grouping as well.

**Comment by marekhrabe** [comment]
I've tried building a super simple experiment with radios:

<a href="https://jsfiddle.net/marek/hrno5w6j/"><img width="125" alt="Screenshot 2020-03-01 at 11 41 45" src="https://user-images.githubusercontent.com/156676/75624121-b3141c00-5bb1-11ea-8c23-5013bb7084c2.png"></a>
https://jsfiddle.net/marek/hrno5w6j/

Arrow keys work (almost) as expected and they definitely feel better than the current implementation where the arrow keys move your cursor in the editor, possibly changing focus to a different block and completely hiding the color picker. I haven't tested yet with an assistive technology but I expect radios to have a more logical structure than separate buttons. 

Advantages:
- radio inputs seem like a logical choice as they are all focused as "one control" (single tab) and further interactions work using arrow keys
- it makes sense semantically as we are selecting one value from a list of things that are all visible

Things I've noticed:
- arrow keys don't move you in two axis but they only move you back/forward in the list of radios in the order they appear in the markup. That means both Left key and Up key do the same - select a previous option and the same is true for Right/Down - moving you forward. I don't mind this much but it could be a surprising interaction.
- there is a feature of the current implementation that clears the color setting if you click a color swatch that is already selected. Standard radio inputs don't work this way. We still have the same clearing feature accessible using the "Clear" button but that's something that 

I don't mind either of these but we need to evaluate it in context of already having an implementation live and we need to look at it from accessibility perspective too. We can most likely code these non-standard interactions on top of radio buttons but I think we might be better sticking with the most standard way possible.

Would you think this could be a good way forward?

**Comment by jeryj** [comment]
Thanks for putting this together, @marekhrabe!

I like the idea of using radio buttons. 

> arrow keys don't move you in two axis but they only move you back/forward in the list of radios in the order they appear in the markup. That means both Left key and Up key do the same - select a previous option and the same is true for Right/Down - moving you forward. I don't mind this much but it could be a surprising interaction.

I agree with you, but my guess is that most people who are not familiar with this arrow interaction are not people who generally use keyboards to interact with the inputs. 

> We can most likely code these non-standard interactions on top of radio buttons but I think we might be better sticking with the most standard way possible.

100% agree. Coding non-standard interactions feels like digging ourselves into a hole. 

My biggest concern with using radio buttons is that the default radio button interaction is to move to _and select_ the next radio button. So, if someone had a custom color set already, then cycling through the radio buttons would clear their selection. Since these don't actually _look_ like radio buttons:
- I don't expect this would be clearly indicated to the user what is happening (their cycling through is selecting the color)
- I'd also be worried that if they _did_ know it was selecting it, they wouldn't actually want that interaction.

I'm going to have to think over this some more... I'm so glad you put together that little interaction and your thoughts on it. It's very, very helpful!

**Comment by marekhrabe** [comment]
> My biggest concern with using radio buttons is that the default radio button interaction is to move to and select the next radio button. So, if someone had a custom color set already, then cycling through the radio buttons would clear their selection. 

This is an important one, thanks for mentioning! I kinda like that cycling through the options provides you with an immediate "preview". What might be concerning is possibly creating way too much undo operations than expected. I believe each arrow key press would leave an undo operation so in order to go back to whatever you started with, you might need to trigger undo for each color selected on the way.

**Comment by ZebulanStanphill** [comment]
Creating a bunch of undo operations would happen with any radio option, though, wouldn't it? You don't need to use the arrow keys anyway unless you want to change the option, as far as I can tell. If you just want to get to the next control on the page, you would just press tab. I would assume there's no reason to use the arrow keys to navigate through radio options other than to select a new option, so if it's consistent with other radio controls I would consider that a plus.

I also think that having to tab through each option (which is the current behavior) before you can get to the next control is not good. At the very least, I think we should change this so that the arrow keys are used for switching focus between options within a color picker, and tab/shift-tab is used for moving in/out of the control as a whole.

**Comment by jeryj** [comment]
> Creating a bunch of undo operations would happen with any radio option, though, wouldn't it? You don't need to use the arrow keys anyway unless you want to change the option, as far as I can tell. If you just want to get to the next control on the page, you would just press tab. I would assume there's no reason to use the arrow keys to navigate through radio options other than to select a new option, so if it's consistent with other radio controls I would consider that a plus.

All excellent points 🙌 

I'm 100% in favor of using whatever the most semantic HTML element for this is and not rebuilding the interactions. If that's a `radio`, then let's give it a try :)

**Comment by enriquesanchez** [comment]
This issue was [discussed today during the Accessibility Team's weekly meeting](https://wordpress.slack.com/archives/C02RP4X03/p1584118088482200).

We found that the color picker component is already using a a <fieldset> and <legend>.

The consensus is that the team has no strong preference over which method to use. We think Radio buttons could work, since they have a similar interaction as the current implementation.

One concern was that with the current implementation, you can press a button to disable the color (and reset to default color), but with radio buttons that’s not possible natively. The only option would be the “Clear” button or add a radio option of "default".

**Comment by paaljoachim** [comment]
We discussed it during the Gutenberg Design Triage. This is really not a design issue but a dev issue.
It would be great if a dev can create a draft PR and test how it works.
